TOP-21 feature signup terms

cwj-c opened the pull request:
## Explain this Pull Request 🙏

- 회원가입 프로세스 중 이용약관 페이지 구현
- 회원가입 프로세스 중 이용약관 내용 페이지 구현

## What has changed? 🤔
- 변경 사항 
  - TermsItemView 구현
  - TermsViewModel 구현
  - TermsActivity 구현
  - TermsModel, TermsEntity 형태 수정
  - TermsContentActivity 구현
  - TermsContentViewModel 구현
  - TermsContentItemView 구현

## Screenshot 📸


## Test Checklist 🚩

cwj-c committed:
feat: TOP-21 TermsEntity, TermsModel에 description 추가

- signup_terms.json 형태에 description을 추가하여 관련된 데이터 모델, 엔티티에도 변경을 적용
- 관련 테스트 케이스 수정

cwj-c committed:
fix: TOP-21 Map을 Json으로 변환할때 Key가 String형태여야 하는 문제 수정

 - Object를 key로 한 Map을 Json으로 변환시, toString()이 호출되어 Key가 String형태로 변환
 - 해당 Json을 Object로 되돌릴때 JsonSyntaxException 발생
 - SignupUserEntity의 termsAgreement Map형태를 String to Pair<TermsModel, Boolean>으로 변경하여 문제 해결

cwj-c committed:
feat: TOP-21 View.getLayoutInflater() extension 메소드 추가

cwj-c committed:
feat: TOP-21 ItemViewBindingDelegate 함수 작성

cwj-c committed:
feat: TOP-21 StringProvider에 이용약관 관련 string resource 추가

cwj-c committed:
feat: TOP-21 TermsItemView 구현

cwj-c committed:
feat: TOP-21 TermsViewModel 구현

cwj-c committed:
feat: TOP-21 TermsActivity 구현

cwj-c committed:
feat: TOP-21 EmailActivity에 TermsActivity 연결

cwj-c committed:
remove: TOP-21 remove unuse code

cwj-c committed:
feat: TOP-21 TermsItemView 클릭 이벤트 범위 조정

- background 클릭 시 select event 전달
- right arrow 클릭 시 click event 전달

cwj-c committed:
feat: TOP-21 TermsActivity 전체선택 글씨를 눌러도 전체선택 기능이 동작하게 설정

cwj-c committed:
feat: TOP-21 TermsActivity 세로 고정 코드 추가

cwj-c committed:
feat: TOP-21 TermsModel을 직렬화 가능하도록 Serializable을 구현하도록 설정

cwj-c committed:
feat: TOP-21 repeatOnCreated 확장함수 생성 후, Terms collect를 해당 확장함수에서 실행

- onStop-onStart 호출 시, AddView가 누적되는 현상 발견
- Terms가 변경될 일이 없으니, 관찰 생명주기를 CREATED로 변경

cwj-c committed:
design: TOP-21 TermsActivity의 layout depth 조절, 스크롤에 tool bar가 포함되지 않도록 수정

cwj-c committed:
Merge branch 'feature/TOP-13_feature_signup' into feature/TOP-21_feature_signup_terms

cwj-c committed:
fix: TOP-21 TermsActivity에 repeatOnCreate import문 커밋 누락

cwj-c committed:
Merge branch 'feature/TOP-21_feature_signup_terms' of https://github.com/THT-Team/THT-Android into feature/TOP-21_feature_signup_terms

cwj-c committed:
feat: TOP-21 TermsContentViewModel 구현

cwj-c committed:
feat: TOP-21 TermsContentItemView 구현

cwj-c committed:
feat: TOP-21 P1SemiBoldTextStyle 추가

cwj-c committed:
feat: TOP-21 TermsContentActivity 구현

cwj-c committed:
feat: TOP-21 TermsActivity에 TermsContentItemView 연결

cwj-c committed:
feat: TOP-21 signup.terms.json 내용 수정

- 스크롤 테스트에 용이하도록 내용의 양을 늘림

cwj-c committed:
fix: TOP-21 drawable resource 이름이 수정됨을 적용

cwj-c committed:
refactor: TOP-21 repeatOnCreate 확장함수의 의미가 없다고 판단하고 제거

cwj-c committed:
design: TOP-21 TermsActivity의 배경색상이 미적용된 부분 적용

cwj-c committed:
fix: TOP-21 TermsContentActivity의 onStart가 호출될떄마다 addView가 누적되는 현상 수정

cwj-c committed:
fix: TOP-21 TermsActivity에 repeatOnCreate import문 제거 누락

kwangddang commented on feature/signin/src/main/res/layout/activity_terms_content.xml:
```diff
@@ -0,0 +1,67 @@
+        <TextView
+            android:layout_gravity="center"
+            style="@style/H5BoldTextStyle"
+            android:textColor="@color/white_f9fafa"
+            android:id="@+id/tv_tool_bar_title"
```
default 색이 white이지 않나요?

cwj-c replied:
default색상이 다른 이유로 변경될수도 있다고 생각이 들어서 색상은 모두 개별적으로 기입을 해놓았습니다..!!

cwj-c committed:
fix: TOP-21 ViewModel의 init에서 phone의 유효성이 검증되지 않을 경우 FinishView SideEffect를 통지하지 않고, InvalidatePhone UiState를 통지하도록 수정

- SideEffect를 observe하기 이전에 통지하는 경우 이벤트 전달이 안되는 문제 수정
- InvalidatePhone 상태에서는 View진행이 되지 않아야 하기에, UiState에 더 적합
- Verify, Email, Terms 3개 화면 수정

cwj-c committed:
rename: TOP-21 Verify의 네이밍을 PhoneVerify로 더 명확한 네이밍을 가지도록 수정

cwj-c committed:
feat: TOP-21 TermsActivity의 NextViewEvent에 SignupRootActivity 연동

kwangddang commented on feature/signin/src/main/java/tht/feature/signin/auth/PhoneVerifyViewModel.kt:
```diff
@@ -47,10 +47,8 @@ class VerifyViewModel @Inject constructor(
-                VerifySideEffect.FinishView(
-                    stringProvider.getString(StringProvider.ResId.InvalidatePhone)
-                )
+            _uiStateFlow.value = VerifyUiState.InvalidatePhone(
+                stringProvider.getString(StringProvider.ResId.InvalidatePhone)
```
지금까지는 toast메시지를 보여주는 걸 SideEffect에서 ShowToast()로 하지 않았나요? 바뀐 이유가 궁금합니다!

cwj-c replied:
SideEffect는 SharedFlow로 관리를 하기때문에 observe를 수행하기 이전에 emit된 데이터들은 누락되는 문제가 있습니다

ViewModel의 init에서 phone값을 체크하여 유효하지 않을경우 FinishView SideEffect를 통지시켜 데이터가 유효하지 않음을 알림과 함께 View를 종료시켰었는데요, ViewModel의 init시점에는 View에서 observe를 진행하지 않은 상태이기 때문에 통지된 SideEffect가 누락됩니다
그렇기에 유효하지 않은 phone데이터를 전달받더라도 View가 종료되지 않게됩니다

때문에 InvalidatePhone이라는 UiState로 변경하여, observe를 수행했을때 현재 View상태가 유효하지 않은 phone데이터를 가지고 있으니, 그에 대한 처리를 진행하라는 의미로 변경을 하였습니다

kwangddang commented on feature/signin/src/main/java/tht/feature/signin/terms/TermsViewModel.kt:
```diff
@@ -0,0 +1,170 @@
+
+    sealed class TermsUiState : UiState {
+        object SelectNone : TermsUiState()
+        data class Select(
+            val selectTermsSet: Set<TermsModel>,
+            val isRequireSelect: Boolean
+        ) : TermsUiState()
```
제가 코드를 보면서 이부분 때문에 조금 헷갈렸었는데 `isRequireSelect`가 전체 체크를 했는지 안 했는지를 나타내주는 변수인가요? 맞다면 `isAllSelected` 같은 네이밍이 조금 더 한눈에 들어오지 않을까 의견 제시해봅니다..!

cwj-c replied:
전체체크를 했는지에 대한 의미가 부분적으로 맞습니다!!

단순히 모든 변수를 체크한것에 대한 정보가 아닌, 동의가 필수적으로 요구되는 값들이 모두 선택되었는지에 대한 정보값이라 해당 네이밍을 사용했습니다

예를들어 동의가 필수적인 약관들이 모두 동의되었고, 동의가 선택적인 약관들이 모두 동의되지 않았을때, `isRequireSelect`는 true가 됩니다

광현님 말씀을 들어보니 조금 의미가 모호했다고 생각이 들어서, `isRequireTermsAllSelect`로 변경했습니다!

kwangddang commented on data/src/main/java/com/tht/tht/data/local/entity/SignupUserEntity.kt:
```diff
@@ -4,7 +4,7 @@ import com.tht.tht.domain.signup.model.TermsModel
 data class SignupUserEntity(
     val phone: String,
-    val termsAgreement: Map<TermsModel, Boolean>,
+    val termsAgreement: Map<String, Pair<TermsModel, Boolean>>,
```
String을 키 값으로 가지는 Map으로 바꾼 이유가 어떤 건가용?

cwj-c replied:
Map의 경우, 직렬화 과정에서 Key값을 String형태로 변경을 합니다
때문에 Key값이 객체라면, `toString()`을 호출한 결과로 직렬화를 수행합니다

이때, 역직렬화를 수행할 경우, String값을 Object로 변환하지못하는 문제가 발생하기때문에 직렬화와 역직렬화를 수행하는 Entity는 Key값을 String으로 변경을 해주는 코드를 추가하였습니다

https://github.com/THT-Team/THT-Android/pull/42/commits/cf7fe047262bd00e24932edf5c03936bf37c4ad2 커밋에서 변경을 하였는데요, 관련 커밋메시지에도 해당 내용 기입해 놓았습니다!!

kwangddang commented on feature/signin/src/main/java/tht/feature/signin/terms/TermsActivity.kt:
```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    private fun addTermsView(terms: TermsModel) {
+        val prevView = (binding.layoutBackground.children.first() as? TermsItemView) ?: binding.ivCheckAll //35
+        val termsItemView = TermsItemView(this).apply {
```
35라는 주석은 일부러 남겨놓으신건가요?

cwj-c replied:
제거되었어야 하는데 누락된 부분이네요
제거하도록 하겠습니다!!

kwangddang reviewed (approved):
고생하셨습니다

cwj-c committed:
Merge branch 'feature/TOP-13_feature_signup' into feature/TOP-21_feature_signup_terms

cwj-c committed:
fix: TOP-21 THTSignupApiService의 requestVeirfy 메소드의 네이밍이 requestPhoneVerify로 변경된것에 대한 적용

cwj-c committed:
remove: TOP-21 remove unuse comment

cwj-c committed:
rename: TOP-21 isRequireSelect 변수를 의미에 걸맞게 isRequireTermsAllSelect로 변경

- 필수 선택 약관들이 모두 선택되었는지에 대한 의미를 더 잘 나타내도록 네이밍 변경
